python/_restclient/hydrotools/_restclient/utilities.py

```diff
@@ -0,0 +1,181 @@
+#!/usr/bin/env python3
+
+from copy import deepcopy
+from dataclasses import dataclass
+from typing import Any, Hashable, Iterable, List, Union
+
+# local import
+from ._iterable_nonstring import IterableNonStringLike
+
+
+@dataclass(frozen=True)
+class Alias:
+    """Create an immutable key, value pair where values alias the key. This is often
+    useful in a variety of applications when the API differs from the backend value.
+    This is also useful when writing factories patterns.
+
+    Any type is a valid key and is deepcopied at construction, meaning keys cannot be
+    mutated by reference. Valid alias values are single atomic types and collections
+    of atomic types. Values are stored in a frozenset, so in the case of passing a
+    dictionary as a value, on the keys from the passed dictionary will be considered.
+
+    Examples:
+        cms = Alias("cms", ["CMS", "m^3/s"])
+
+        cms["CMS"] # returns "cms"
+        cms.get("m^3/s") # returns "cms"
+        "CMS" in cms # returns True
+
+        # foo is callable
+        foo_alias = Alias(foo, ["bar", "baz"])
+
+        result = foo_alias["bar"]()
+    """
+
+    key: Any
+    valid_value: Union[Iterable, Hashable]
+
+    def __post_init__(self):
+        # Create deep copy so a ref to a mutable key could not change value implicitly
+        self.__dict__["key"] = deepcopy(self.key)
```

Comment:
The assumption that `Any` type is `deepcopy`able may not be appropriate.  Especially for custom and/or extension types.

Reply:
Agreed, having done some reading to see if I could figure out a way to determine if an object will be `deepcopy`able at construction and [this](https://stackoverflow.com/a/18611983) is all that ive found. It looks like an extension type would have to be pickable for it to be `deepcopy`able. Unfortunately I havent found an easy way to check if an object is pickable or not without just pickling the object itself. This seems like unnecessary overhead, so as we talked about offline I think the best course of action is to only accept Callables or Scalar types here.